contracts/UniswapV3Pool.sol

```diff
@@ -139,13 +139,23 @@ contract UniswapV3Pool is IUniswapV3Pool, NoDelegateCall {
     }
 
     /// @dev Get the pool's balance of token0
+    /// @dev This function is gas optimized to avoid a redundant extcodesize check in addition to the returndatasize
+    /// check
     function balance0() private view returns (uint256) {
-        return IERC20Minimal(token0).balanceOf(address(this));
+        (bool success, bytes memory data) =
+            token0.staticcall(abi.encodeWithSelector(IERC20Minimal.balanceOf.selector, address(this)));
+        require(success && data.length == 32);
+        return abi.decode(data, (uint256));
     }
 
     /// @dev Get the pool's balance of token1
+    /// @dev This function is gas optimized to avoid a redundant extcodesize check in addition to the returndatasize
+    /// check
     function balance1() private view returns (uint256) {
-        return IERC20Minimal(token1).balanceOf(address(this));
+        (bool success, bytes memory data) =
+            token1.staticcall(abi.encodeWithSelector(IERC20Minimal.balanceOf.selector, address(this)));
```

Comment:
Hey @moodysalem 
I'm interested in why you decided to it this way? (change from token.balanceOf)
Why not validate that the tokens are contracts in the constructor, and if somebody potentially selfdestructs the token, eventually transfer call will revert?

Reply:
if you look at gas snapshots it saves ~400 gas. probably because old versions of solidity would do extcodesize checks for calls that returned data (not entirely sure that's what saved gas here)

Reply:
Thanks for the quick answer 🍻